web/packages/teleterm/src/ui/Search/pickers/ActionPicker.tsx

```diff
@@ -245,6 +243,109 @@ export const InputWrapper = styled(Flex).attrs({ px: 2 })`
   }
 `;
 
+const ExtraTopComponents = (props: {
+  status: ActionPickerStatus;
+  getClusterName: (resourceUri: uri.ClusterOrResourceUri) => string;
+  showErrorsInModal: (errors: ResourceSearchError[]) => void;
+}) => {
+  const { status, getClusterName, showErrorsInModal } = props;
+
+  switch (status.status) {
+    case 'no-input': {
+      return status.hasSelectedAllFilters && <TypeToSearchItem />;
+    }
+    case 'processing': {
+      return null;
+    }
+    case 'finished': {
+      return (
+        <>
+          {status.nonRetryableResourceSearchErrors.length > 0 && (
+            <ResourceSearchErrorsItem
+              errors={status.nonRetryableResourceSearchErrors}
+              getClusterName={getClusterName}
+              showErrorsInModal={() => {
+                showErrorsInModal(status.nonRetryableResourceSearchErrors);
+              }}
+            />
+          )}
+          {status.hasNoResults && (
+            <NoResultsItem
+              clustersWithExpiredCerts={status.clustersWithExpiredCerts}
+              getClusterName={getClusterName}
+            />
+          )}
+        </>
+      );
+    }
+    default: {
+      assertUnreachable(status);
+    }
+  }
+};
+
+type ActionPickerStatus =
+  | { status: 'no-input'; hasSelectedAllFilters: boolean }
+  | { status: 'processing' }
+  | {
+      status: 'finished';
+      hasNoResults: boolean;
+      nonRetryableResourceSearchErrors: ResourceSearchError[];
+      clustersWithExpiredCerts: Set<uri.ClusterUri>;
+    };
```

Comment:
When making `ActionPickerStatus`, I tried to follow the principle of [making sure that impossible states are impossible to express](https://www.youtube.com/watch?v=IcgmSRJHu_8) with data structures. Initially some properties were included directly in the `ActionPickerState` while others were gated behind a discriminated union.

For example, I knew that `hasNoResults` would make no sense if `status` was different than `finished`, because if the search hasn't finished yet how we could be sure that there's no results.

But there are fields which technically aren't tied to the status. `hasSelectedAllFilters` can be true or false no matter whether the search is in progress or not. `clustersWithExpiredCerts` could technically include only clusters from `ClustersService` when the search is still in progress.

Initially I included those directly on the object. However, soon I refactored the code to include them only where necessary and tied all of them to particular statuses. When reading the type alone, it's more clear in what situation the given property is used. If there was a field used by 2/3 of the statuses, then I'd probably repeat it for those two statuses.